org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/editor/WRecordIDEditor.java

```diff
@@ -0,0 +1,367 @@
+		boolean isSOTrx = true;
+		if(po.columnExists("IsSOTrx")) {
+			isSOTrx = po.get_ValueAsBoolean("IsSOTrx");
+		}
+		int AD_Window_ID;
+		if(isSOTrx)
+			AD_Window_ID = savedTable.getAD_Window_ID();
+		else
+			AD_Window_ID = savedTable.getPO_Window_ID();
+
+		if (AD_Window_ID > 0) {
+			if(!Util.isEmpty(keyColumn)) {
+				MQuery zoomQuery = new MQuery();
+				zoomQuery.addRestriction(keyColumn, MQuery.EQUAL, (Integer)po.get_Value(keyColumn));
+				zoomQuery.setRecordCount(1);
+	
+				AEnv.zoom(AD_Window_ID, zoomQuery);
+			}
+		}
+	}
+
+	@Override
+	public void onMenu(ContextMenuEvent evt)
+	{
+		if (WEditorPopupMenu.REQUERY_EVENT.equals(evt.getContextEvent()))
+		{
+			actionRefresh(oldValue);
+		}
+		else if (WEditorPopupMenu.ZOOM_EVENT.equals(evt.getContextEvent()))
+		{
+			actionZoom();
+		}
+		else if (WEditorPopupMenu.CHANGE_LOG_EVENT.equals(evt.getContextEvent()))
+		{
+			WFieldRecordInfo.start(gridField);
+		} 
+		else if (WEditorPopupMenu.PREFERENCE_EVENT.equals(evt.getContextEvent()))
+		{
+			if (isShowPreference())
+				ValuePreference.start (getComponent(), gridField, getValue());
+		}
+	}
+
+	@Override
+	public void setReadWrite(boolean readWrite) {
+		m_ReadWrite = readWrite;
+		if(editButton != null)
+			editButton.setDisabled(!readWrite);
+		
+	}
+
+	@Override
+	public boolean isReadWrite() {
+		return m_ReadWrite;
+	}
+
+	public  void propertyChange(PropertyChangeEvent evt)
+	{
+		if (GridField.PROPERTY.equals(evt.getPropertyName()))
+		{
+			setValue(evt.getNewValue(), false);
+		}
+	}
+	
+	@Override
+	public void setValue(Object value) {
+		setValue(value, true);
+	}
+	
+	/**
+	 * 	Set Value
+	 *	@param value value
+	 *	@param fire data binding
+	 */
+	
+	private void setValue (Object value, boolean fire) {
+		if (value == null || value.toString().trim().length() == 0)
+		{
+			oldValue = null;
+			value = null;
+			recordTextBox.setValue("");
+		}
+		else {
+			if(oldTableID != tableIDGridField.getValue()) {
+				oldTableID = tableIDGridField.getValue();
+			}
+			if(value != null && oldTableID != null) {
+				Integer recordID = Integer.parseInt(String.valueOf(value));
+				Integer tableID = Integer.parseInt(String.valueOf(oldTableID));
+				if(recordID != null && tableID != null)
+					recordTextBox.setValue(getIdentifierColumns(tableID, recordID));
+			}
+		}
+
+//		Data Binding
+		if (fire) {
+			ValueChangeEvent changeEvent = new ValueChangeEvent(this, this.getColumnName(), oldValue, value);
+			super.fireValueChange(changeEvent);
+		}
+		oldValue = value;
+
+	}
+
+	@Override
+	public Object getValue() {
+		return oldValue;
+	}
+
+	@Override
+	public String getDisplay() {
+		if(((Integer) tableIDGridField.getValue() != null) && ((Integer) gridField.getValue() != null))
+			return getIdentifierColumns((Integer) tableIDGridField.getValue(), (Integer) gridField.getValue());
+		else
+			return "";
+	}
+	
+	@Override
+    public String getDisplayTextForGridView(Object value) {
```

Comment:
This wouldn't work for grid view since you can't get the row's table id value. A possible approach to fix this is to add GridRowCtx parameter and you can get the row's table id value with `String key = gridTab.getWindowNo + "|AD_Table_ID"; Object rowTableIdValue = gridRowCtx.get(key);`